pydantic/schema.py

```diff
@@ -346,6 +347,7 @@ def get_model_name_map(unique_models: Set[Type['BaseModel']]) -> Dict[Type['Base
     conflicting_names: Set[str] = set()
     for model in unique_models:
         model_name = model.__name__
+        model_name = re.sub(r"[^a-zA-Z0-9.\-_]", "_", model_name)
```

Comment:
@samuelcolvin I expect this line may need to be removed / changed before merging, but adding this regex substitution fixes all problems with generic class names for `openapi-generator` (and I imagine for most json-schema-based tools). As far as I can tell, it isn't possible to generate a name that would be modified by this regex substitution in the normal course of pydantic usage (i.e., inheriting from `BaseModel`), including them is a violation of the open api spec (and presumably others), and I am having a hard time imagining a scenario where you would want those characters in your model name anyway.

From what I can tell, this substitution would also have little-to-no impact on performance given it's only involved in schema generation and is a fairly cheap regex.

I'm reluctant to change the naming scheme for generic classes because the `repr` and error messages are so clear. At the same time, I think json schemas are one of the most natural contexts for the use of generics (and my intended use is specifically with openapi / `openapi-generator`), so I think it would be unfortunate if they were slightly broken for this use case.

I'd be interested in your thoughts.

Reply:
LGTM

This function is mainly/only used to generate the JSON Schemas (and it's reused by FastAPI).

I think it would be safe/ideal to apply this. And I see you are using the same regex in the OpenAPI spec :heavy_check_mark: